core/src/main/java/org/infinispan/factories/impl/MBeanMetadata.java

```diff
@@ -169,6 +205,31 @@ public Map<String, String> tags() {
          return tags;
       }
 
+      public TimeUnit timeUnit() {
+         return timeUnit;
+      }
+
+      public AttributeMetricType metricType() {
+         if (getterFunction != null) {
+            return AttributeMetricType.GAUGE;
+         }
+         if (counterFunction != null) {
+            return AttributeMetricType.COUNTER;
+         }
+         if (functionTimerFunction != null) {
+            return AttributeMetricType.FUNCTION_TIMER;
+         }
+
+         // histogram based metrics
+         if (setterFunction != null) {
+            return AttributeMetricType.TIMER;
+         }
+         if (summaryFunction != null) {
+            return AttributeMetricType.SUMMARY;
+         }
+         return AttributeMetricType.UNKNOWN;
```

Comment:
This is starting to get pretty convoluted.  Does it make sense to introduce different classes for each metric type? `GaugeMetadata`, `CounterMetadata` etc

Reply:
So instead of having a generic `AttributeMetadata` class with a `metricType()` method, we utilise inheritance to create specific classes for the different types and set the constant fields in the constructor. e.g.:

```java
   public static class GuageMetadata extends AttributeMetadata {
      private final Function<?, ?> getterFunction;  // optional
      
      public GuageMetadata(String name, String description,
                                                   Function<C, Number> function,
                                                   Map<String, String> tags) {
         super(name, description, false, false, null, false)
         this.getterFunction = function;
      }
   }
   
   public static abstract class AttributeMetadata {
      private final String name;
      private final String description;
      private final boolean writable;
      private final boolean useSetter;
      private final String type;
      private final boolean is;
...
   }
```

`MetricUtils` then becomes much simpler:

```java
   public static <C> AttributeMetadata createGauge(String name, String description,
                                                   Function<C, Number> function,
                                                   Map<String, String> tags) {
      Objects.requireNonNull(function);
      return new GaugeMetadata(name, description, function, tags);
   }

```
vs
```java
   public static <C> AttributeMetadata createGauge(String name, String description,
                                                   Function<C, Number> function,
                                                   Map<String, String> tags) {
      Objects.requireNonNull(function);
      return new AttributeMetadata(name, description, false, false, null, false, function, null, null, null, null, false, tags, null);
   }

```

Reply:
and replace the enum with an if chain of `instanceOf`

Reply:
Yeah .. not pretty, but IMO it's clearer than a generic `AttributeMetadata` object with optional fields.